### ISSUE TITLE
Debug the Slow Full Finetuning Issue with Linear Attention

### ISSUE THREAD

**Issue by Malikeh97** [Malikeh97/lingua#12]
After switching to FI, the main preferred approach is to try training the decoder model from scratch, but if the speed is lower in the linear attention setting, it doesn't make sense to switch to that. Good to look into the issue!

**Comment by Malikeh97** [comment]
Made these changes to speed up Kimi Linear: 

<img width="1067" height="874" alt="Image" src="https://github.com/user-attachments/assets/42487e56-7505-4286-a386-1484da25b853" />

**Comment by Malikeh97** [comment]
wandb: Run summary (After Fix):
wandb:                  epoch 2
wandb:              epoch_pct 100
wandb:     model/total_params 1959157088
wandb: model/trainable_params 1959157088
wandb:       train/epoch_loss 0.4886
wandb:             train/loss 0.40625
wandb:             train/step 10890
wandb:        val/exact_match 80.4
wandb:                 val/f1 86.9285
wandb:               val/loss 0.44563

Training speed/throughput: ~1-1.15 it/s


Before Fix:
Epoch 1: EM: 74.20% | F1: 83.15%
Epoch 2: EM: 78.40% | F1: 86.66%
Training speed/throughput: ~4–4.5 it/s

EM increased 2% and sped up 3-4 times in the new setup


Softmax full attention:
Epoch 1: EM: 68.60%, F1: 79.39%
Epoch 2: EM: 73.60%, F1: 84.02%
Training speed/throughput: ~2-2.5 it/s


EM increased ~5% compared to softmax and 1.5/2x faster than softmax

**Comment by Malikeh97** [comment]
<img width="919" height="321" alt="Image" src="https://github.com/user-attachments/assets/a89917eb-56c6-4133-9dd4-5bd017693159" />

**Comment by Malikeh97** [comment]
These are the results for from scratch training: 

<img width="593" height="528" alt="Image" src="https://github.com/user-attachments/assets/77a271d0-6d57-467a-bd2b-bc27010b16fc" />

Key Takeaways

  Accuracy:
  - Softmax consistently outperforms linear attention at every depth.
  - The gap widens dramatically with depth: at 6L the EM gap is small (1.8 pts), but by 22L it's catastrophic — linear attention collapses to
  55.4% EM vs. 85.6% for softmax. The 22L linear model's train loss (2.1) suggests it barely converged at all.
  - Within softmax models, deeper = better (6L→12L→22L gains ~3 pts F1 per jump).

  Training Speed:
  - Counter-intuitively, linear attention is slower than softmax at all scales. At 22L it takes ~42% longer per epoch (60 min vs. 42 min). This is likely due to the KDA linear attention implementation overhead outweighing the theoretical O(n) vs. O(n²) advantage at these sequence lengths.
